airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/util/CoroutineUtils.kt

```diff
@@ -10,11 +10,29 @@ interface CloseableCoroutine {
     suspend fun close()
 }
 
-suspend fun <T : CloseableCoroutine, R> T.use(block: suspend (T) -> R) =
+// this is taken almost verbatim from kotlin's AutoCloseable?.use implementation,
+// but with `suspend` modifiers added.
+suspend inline fun <T : CloseableCoroutine, R> T.use(block: (T) -> R): R {
+    var exception: Throwable? = null
     try {
-        block(this)
+        return block(this)
+    } catch (e: Throwable) {
+        exception = e
+        throw e
     } finally {
-        close()
+        this.closeFinally(exception)
+    }
+}
+
+suspend inline fun CloseableCoroutine.closeFinally(cause: Throwable?): Unit =
+    when {
+        cause == null -> close()
+        else ->
+            try {
+                close()
+            } catch (closeException: Throwable) {
+                cause.addSuppressed(closeException)
+            }
```

Comment:
I'm not completely convinced mixing the two types of continuations `try / catch` and `suspend` will work the way we want, but I think we gotta try it out. 🤞 this just works.